model/src/test/java/com/epam/esm/model/service/impl/GiftCertificateServiceImplTest.java

```diff
@@ -0,0 +1,223 @@
+        CertificateDTO actual = service.add(certificateDTO);
+        assertEquals(actual, certificateDTO);
+    }
+
+    @Test
+    void testAddNegative() throws DaoException, ServiceException {
+        lenient().when(dao.create(any(GiftCertificate.class))).thenReturn(certificate);
+        CertificateDTO actual = service.add(certificateDTO);
+        assertNotEquals(actual, new CertificateDTO());
+    }
+
+    @Test
+    void testAddException() throws DaoException {
+        lenient().when(dao.create(any(GiftCertificate.class))).thenThrow(DaoException.class);
+        assertThrows(ServiceException.class, () -> service.add(certificateDTO));
+    }
+
+    @Test
+    void testUpdatePositive() throws DaoException, ServiceException {
+        lenient().when(dao.update(any(CertificateDTO.class))).thenReturn(1L);
+        lenient().when((dao.read(anyLong()))).thenReturn(certificate);
+        CertificateDTO actual = service.update(certificateDTO);
+        assertEquals(actual, certificateDTO);
+    }
+
+    @Test
+    void testUpdateNegative() throws DaoException, ServiceException {
+        lenient().when(dao.update(any(CertificateDTO.class))).thenReturn(1L);
+        lenient().when((dao.read(anyLong()))).thenReturn(certificate);
+        CertificateDTO actual = service.update(certificateDTO);
+        assertNotEquals(actual, new CertificateDTO());
+    }
+
+    @Test
+    void testUpdateException() throws DaoException {
+        lenient().when(dao.update(any(CertificateDTO.class))).thenThrow(DaoException.class);
+        assertThrows(ServiceException.class, () -> service.update(certificateDTO));
+    }
+
+    @Test
+    void testDeletePositive() throws DaoException, ServiceException {
+        lenient().when(dao.delete(anyLong())).thenReturn(1L);
+        long actual = service.delete(1L);
+        assertEquals(actual, 1L);
+    }
+
+    @Test
+    void testDeleteNegative() throws DaoException, ServiceException {
+        lenient().when(dao.delete(anyLong())).thenReturn(1L);
+        long actual = service.delete(1L);
+        assertNotEquals(actual, 2L);
+    }
+
+    @Test
+    void testDeleteException() throws DaoException {
+        lenient().when(dao.delete(anyLong())).thenThrow(DaoException.class);
+        assertThrows(ServiceException.class, () -> service.delete(1L));
+    }
+
+    @Test
+    void testFindByTagPositive() throws DaoException, ServiceException {
+        lenient().when(dao.findByTag(anyString())).thenReturn(new ArrayList<>());
+        List<CertificateDTO> actual = service.findByTag("tagName");
+        assertEquals(actual, new ArrayList<>());
+    }
+
+    @Test
+    void testFindByTagNegative() throws DaoException, ServiceException {
+        ArrayList<GiftCertificate> list = new ArrayList<>();
+        list.add(certificate);
+        lenient().when(dao.findByTag(anyString())).thenReturn(list);
+        List<CertificateDTO> actual = service.findByTag("tagName");
+        assertNotEquals(actual, new ArrayList<>());
```

Comment:
Probably better to compare it by size of list or using .isEmpty method

Reply:
fix